### ISSUE TITLE
To find predicted FP, FN and TP bounding boxes.

### ISSUE THREAD

**Issue by beenajambucha1-design** [ultralytics/ultralytics#24859]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

i want to find out all the FP, FN and TP bounding boxes which model have used to calculate P, R and F1 score because i want to analyze those objects which model has declared as FP or FN.  By applying visualization = True, it only gives images data from confusion matrix. I used confidence = 0.001 and iou =0.45 and following code to find out but i does not gives values which terminal output gives. Are confidence and iou values varies throughout the validation?
import numpy as np
import torch
# Define IoU threshold and Confidence threshold
iou_threshold = 0.45
confidence_threshold = 0.001 # Added confidence threshold

def xywhn2xyxy(x, w=640, h=640, padw=0, padh=0):
    # Convert nx4 boxes from [x, y, w, h] normalized to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
    # x = [x_center, y_center, width, height] normalized
    y = x.copy() * 1.0
    y[0] = w * (x[0] - x[2] / 2) + padw  # x1
    y[1] = h * (x[1] - x[3] / 2) + padh  # y1
    y[2] = w * (x[0] + x[2] / 2) + padw  # x2
    y[3] = h * (x[1] + x[3] / 2) + padh  # y2
    return y.astype(int)

def box_iou(box1: torch.Tensor, box2: torch.Tensor, eps: float = 1e-7) -> torch.Tensor:

  (a1, a2), (b1, b2) = box1.float().unsqueeze(1).chunk(2, 2), box2.float().unsqueeze(0).chunk(2, 2)
  inter = (torch.min(a2, b2) - torch.max(a1, b1)).clamp_(0).prod(2)

    # IoU = inter / (area1 + area2 - inter)
  return inter / ((a2 - a1).prod(2) + (b2 - b1).prod(2) - inter + eps)

classified_detections = {}

for i, img_path in enumerate(selected_images):
    img_height, img_width = results[i].orig_shape # Get original image dimensions

    # Get ground truth annotations for the current image
    gt_boxes_yolo = ground_truth_annotations.get(img_path, [])
    gt_boxes_abs = []
    for gt_anno in gt_boxes_yolo:
        gt_class_id = gt_anno['class_id']
        gt_bbox_norm = np.array(gt_anno['bbox']) # normalized [x_center, y_center, width, height]
        gt_bbox_abs = xywhn2xyxy(gt_bbox_norm, w=img_width, h=img_height) # absolute [x1, y1, x2, y2]
        gt_boxes_abs.append({'class_id': gt_class_id, 'bbox_abs': gt_bbox_abs, 'matched': False})

    # Get model predictions for the current image
    pred_boxes = results[i].boxes.cpu().numpy() # Boxes object from ultralytics results
    preds_abs = []
    # Corrected: Access shape directly from the Boxes object's data attribute or xyxy attribute
    if pred_boxes.xyxy.shape[0] > 0:
        for j in range(pred_boxes.xyxy.shape[0]):
            if pred_boxes.conf[j] >= confidence_threshold: # Filter by confidence threshold
                pred_bbox_xyxy = pred_boxes.xyxy[j] # absolute [x1, y1, x2, y2]
                pred_class_id = int(pred_boxes.cls[j])
                pred_confidence = pred_boxes.conf[j]
                preds_abs.append({'class_id': pred_class_id, 'bbox_abs': pred_bbox_xyxy, 'confidence': pred_confidence, 'matched': False})

    # Initialize lists for this image
    true_positives = []
    false_positives = []
    false_negatives = []

    # Compare ground truth with predictions
    for gt_idx, gt in enumerate(gt_boxes_abs):
        best_iou = 0
        best_pred_idx = -1

        for pred_idx, pred in enumerate(preds_abs):
            if not pred['matched'] and gt['class_id'] == pred['class_id']:
                # Convert numpy arrays to torch tensors and add a batch dimension
                gt_bbox_tensor = torch.from_numpy(gt['bbox_abs']).float().unsqueeze(0)
                pred_bbox_tensor = torch.from_numpy(pred['bbox_abs']).float().unsqueeze(0)
                current_iou = box_iou(gt_bbox_tensor, pred_bbox_tensor)
                if current_iou > best_iou:
                    best_iou = current_iou
                    best_pred_idx = pred_idx

        if best_iou > iou_threshold:
            true_positives.append({'gt_box': gt, 'pred_box': preds_abs[best_pred_idx]})
            gt_boxes_abs[gt_idx]['matched'] = True
            preds_abs[best_pred_idx]['matched'] = True
        else:
            false_negatives.append(gt)

    # Any remaining unmatched predictions are False Positives
    for pred in preds_abs:
        if not pred['matched']:
            false_positives.append(pred)

    classified_detections[img_path] = {
        'TP': true_positives,
        'FP': false_positives,
        'FN': false_negatives,
        'image_dims': (img_width, img_height)
    }

print(f"Classified detections (TP, FP, FN) for {len(classified_detections)} images.")

total_initial_tp = 0
total_initial_fp = 0
total_initial_fn = 0

for img_path, detection_data in classified_detections.items():
    total_initial_tp += len(detection_data.get('TP', []))
    total_initial_fp += len(detection_data.get('FP', []))
    total_initial_fn += len(detection_data.get('FN', []))

print(f"\n--- Overall Initial TP, FP, FN Counts (before reclassification) ---")
print(f"Total True Positives: {total_initial_tp}")
print(f"Total False Positives: {total_initial_fp}")
print(f"Total False Negatives: {total_initial_fn}")

# Calculate overall metrics
overall_precision_initial = total_initial_tp / (total_initial_tp + total_initial_fp) if (total_initial_tp + total_initial_fp) > 0 else 0.0
overall_recall_initial = total_initial_tp / (total_initial_tp + total_initial_fn) if (total_initial_tp + total_initial_fn) > 0 else 0.0
overall_f1_score_initial = 2 * (overall_precision_initial * overall_recall_initial) / (overall_precision_initial + overall_recall_initial) if (overall_precision_initial + overall_recall_initial) > 0 else 0.0

print(f"\n--- Overall Metrics (before reclassification) ---")
print(f"Precision: {overall_precision_initial:.4f}")
print(f"Recall: {overall_recall_initial:.4f}")
print(f"F1 Score: {overall_f1_score_initial:.4f}")

# Calculate class-wise metrics
class_initial_tp = {i: 0 for i in range(len(class_names))}
class_initial_fp = {i: 0 for i in range(len(class_names))}
class_initial_fn = {i: 0 for i in range(len(class_names))}

for img_path, detection_data in classified_detections.items():
    for tp in detection_data.get('TP', []):
        class_initial_tp[tp['gt_box']['class_id']] += 1
    for fp in detection_data.get('FP', []):
        class_initial_fp[fp['class_id']] += 1
    for fn in detection_data.get('FN', []):
        class_initial_fn[fn['class_id']] += 1

print(f"\n--- Class-wise Metrics (before reclassification) ---")
for class_id, class_name in enumerate(class_names):
    tp = class_initial_tp[class_id]
    fp = class_initial_fp[class_id]
    fn = class_initial_fn[class_id]

    precision = tp / (tp + fp) if (tp + fp) > 0 else 0.0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0.0
    f1_score = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0.0

    print(f"\nClass: {class_name} (ID: {class_id})")
    print(f"  True Positives: {tp}")
    print(f"  False Positives: {fp}")
    print(f"  False Negatives: {fn}")
    print(f"  Precision: {precision:.4f}")
    print(f"  Recall: {recall:.4f}")
    print(f"  F1 Score: {f1_score:.4f}")

**Comment by UltralyticsAssistant** [comment]
👋 Hello @beenajambucha1-design, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue moving quickly, and an Ultralytics engineer will also assist soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
Since this appears to be a custom validation/analysis ❓ question about TP, FP, and FN bounding boxes, please provide as much information as possible, including:
- the exact `ultralytics` version you are using
- the full validation command or script
- a small sample dataset or a simplified script showing the mismatch
- relevant validation logs and terminal output
- a few dataset image examples with corresponding labels
- confirmation that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips)
If this is actually an unexpected behavior or 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Yes, the mismatch is expected: validation `P/R/F1` are produced by [`ap_per_class()`](https://docs.ultralytics.com/reference/utils/metrics#ultralytics.utils.metrics.ap_per_class) by sweeping confidence over all detections and using IoU 0.5 at the max-F1 operating point, while mAP uses IoU 0.5:0.95 and [`ConfusionMatrix.process_batch()`](https://docs.ultralytics.com/reference/utils/metrics#ultralytics.utils.metrics.ConfusionMatrix.process_batch) uses a fixed matcher `iou_thres` (with default detect `conf=0.001` internally mapped to `0.25`), so a post-hoc fixed `conf/iou` pass on `results[i].boxes` will not reproduce the terminal values; for the exact TP/FP/FN boxes, use the validator’s stored confusion-matrix matches rather than recomputing them manually.
